x/wasm/keeper/keeper.go

```diff
@@ -1213,3 +1218,37 @@ func (h DefaultWasmVMContractResponseHandler) Handle(ctx sdk.Context, contractAd
 	}
 	return result, nil
 }
+
+// PruneWasmCodes deletes code info for unpinned codes smaller than given code id.
+func (k Keeper) PruneWasmCodes(ctx sdk.Context, codeID uint64) error {
+	pinnedCodesHash := make(map[string]struct{}, 0)
+	unpinnedCodes := make([]uint64, 0)
+
+	for c := uint64(1); c < codeID; c++ {
+		info := k.GetCodeInfo(ctx, c)
+		if info == nil {
+			continue
+		}
+
+		if k.IsPinnedCode(ctx, c) {
+			pinnedCodesHash[hex.Dump(info.CodeHash)] = struct{}{}
+			continue
+		}
+		unpinnedCodes = append(unpinnedCodes, c)
+	}
+
+	for _, code := range unpinnedCodes {
+		info := k.GetCodeInfo(ctx, code)
+		if info == nil {
+			continue
+		}
+
+		if _, found := pinnedCodesHash[hex.Dump(info.CodeHash)]; found {
+			continue
+		}
+
+		k.deleteCodeInfo(ctx, code)
+		_ = k.wasmVM.RemoveCode(info.CodeHash)
```

Comment:
Do we really want to ignore those errors? It typically means a file could not be removed and indicates invalid state.

Reply:
It is either abort the whole operation or not. IMHO there is value in removing the code-info data from the chain state even in case of an error here. There may be individual reasons on single machines, like file permissions that must not affect consensus. On the next state sync, the file would not be transferred to new nodes.
Logging the error makes sense IMHO